### DISCUSSION TITLE
Is require'ing lua modules outside of plugin specs bad for performance?

### DISCUSSION THREAD

**Discussion by dotfrag** [LazyVim/LazyVim#3829] [Q&A] [answered]
The following two code blocks appear to do the same:

```lua
local biome = require("lspconfig.server_configurations.biome")
local biome_ft = biome.default_config.filetypes
biome_ft = vim.tbl_filter(function(ft)
  return not vim.tbl_contains({ "astro" }, ft)
end, biome_ft)

return {
  "neovim/nvim-lspconfig",
  opts = {
    servers = {
      biome = { filetypes = biome_ft },
    },
  },
}
```

```lua
return {
	"neovim/nvim-lspconfig",
	opts = function(_, opts)
		local biome = require("lspconfig.server_configurations.biome")
		local biome_ft = biome.default_config.filetypes
		biome_ft = vim.tbl_filter(function(ft)
			return not vim.tbl_contains({ "astro" }, ft)
		end, biome_ft)

		opts.servers.biome = {
			biome = { filetypes = biome_ft },
		}
	end,
}
```

Is one preferred over the other? Does the first one decrease performance by potentially requiring a plugin before it's loaded? Or does it not matter, because the performance hit happens on `setup()`, rather than `require`?

**Comment by dpetka2001** [comment]
I'm really no expert on this. Just sharing my experience regarding this. 

In my old config when I used Telescope, I had some functions (to enable multi-selection on entries, maybe that's not needed any more in Telescope, but back then it was not supported out of the box) outside of the returned spec and those were calculated during the parsing phase of `lazy.nvim` and that lead to Telescope having about 16ms in the Profile tab of Lazy UI. When I moved the functions inside the `opts` function Telescope dropped to around 0.15ms. But I did observe a slight delay when I invoked Telescope via `keys`. I think when you have `require`s before the returned spec, it definitely adds up to startup time. It also depends on what you `require` and what you do.

So, in the end I believe it's just a matter of where you will transfer the extra overhead. Will it be in the Neovim startup time or in the plugin loading time when you lazy-load it? I'm not sure myself what the best practice would be.

**Reply by dotfrag** [reply]
Yes, you seem to be right.. I'm noticing similar behavior when testing that. Now I'm debating what I want :smiling_face_with_tear: .

**Reply by dpetka2001** [reply]
I personally chose the latter. Transferring the overhead to the plugin loading and keeping Neovim startup time minimum. But I don't use Telescope any more, so don't have to deal with it any more :stuck_out_tongue:

**Comment by folke** [comment, marked as answer]
The first option would error when you have no plugins installed, and also prevent any kind of lazy loading. So def second option. Always use closures

**Reply by dotfrag** [reply]
Thank you both for the input! I'll accept this answer since it's more definitive.

Something like [this](https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/plugins/extras/lang/java.lua) seems fine because it's only a function definition, but what about something like [this](https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/plugins/extras/editor/dial.lua) (with require statement), or [this](https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/plugins/extras/editor/refactoring.lua)?

If understand correctly, all of the above are okay, because they are functions and will be executed only when called?

**Reply by dpetka2001** [reply]
Well these `require`s are inside a function already that will be called when necessary like you mentioned, so that makes them already a closure. The problem is when you directly `require` at the top-level, like you do in your first example `local biome = require("lspconfig.server_configurations.biome")`.

**Reply by dotfrag** [reply]
Got it, thanks!
